packages/core/src/shared/clients/chatDb/helpers.ts

```diff
@@ -0,0 +1,211 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import fs from '../../fs/fs'
+import path from 'path'
+
+import { TabType } from '../../../amazonq/webview/ui/storages/tabsStorage'
+import { ChatItem, ChatItemButton, DetailedListItem, DetailedListItemGroup, MynahIconsType } from '@aws/mynah-ui'
+
+export const TabCollection = 'tabs'
+
+export const historyPath = path.join('.aws', 'amazonq', 'history')
+
+export type Tab = {
+    historyId: string
+    isOpen: boolean
+    updatedAt: Date
+    tabType: TabType
+    title: string
+    conversations: Conversation[]
+}
+
+export type Conversation = {
+    conversationId: string
+    clientType: ClientType
+    messages: ChatItem[]
+}
+
+export enum ClientType {
+    VSCode = 'IDE-VSCode',
+    JetBrains = 'IDE_JetBrains',
+    CLI = 'CLI',
+}
+
+export class FileSystemAdapter {
+    private directory
+    constructor(directory: string) {
+        this.directory = directory
+    }
+
+    async ensureDirectory() {
+        try {
+            await fs.exists(this.directory)
+        } catch {
+            await fs.mkdir(this.directory)
+        }
+    }
+
+    async loadDatabase(dbname: string, callback: (data: string | undefined | Error) => void) {
+        try {
+            await this.ensureDirectory()
+            const filename = path.join(this.directory, dbname)
+
+            const exists = await fs.exists(filename)
+            if (!exists) {
+                callback(undefined)
+                return
+            }
+
+            const data = await fs.readFileText(filename)
+            callback(data)
+        } catch (err: any) {
+            callback(err)
+        }
+    }
+
+    async saveDatabase(dbname: string, dbstring: string, callback: (err: Error | undefined) => void) {
+        try {
+            await this.ensureDirectory()
+            const filename = path.join(this.directory, dbname)
+
+            await fs.writeFile(filename, dbstring, 'utf8')
```

Comment:
what happens if the history is corrupted or has partial writes, or multiple vscode instances try to save the file simultaneously? on Windows, this will have problems.

if you are ok with the current limitations, the limitations should be cleary called-out in the docstring for `FileSystemAdapter` (which currently has no docstring).